Thread panic in FastAPI streaming response

I have a FastAPI app that returns a streaming response. I instantiate the app via this CLI command:

```
granian --interface asgi --host 0.0.0.0 --port 8000 --reload --log-level debug --access-log factory:APP
```

Irregularly (i.e., once every 2-3 tries), I get the following stacktrace:

```
thread '<unnamed>' panicked at /rustc/129f3b9964af4d4a709d1383930ade12dfe7c081/library/core/src/ptr/mod.rs:514:1:
Cannot drop pointer into Python heap without the GIL being held.
stack backtrace:
[23:29:31] info: [2024-07-08 23:29:29 +0000] 192.168.65.1 - "POST /api/v1/stream/ HTTP/1.1" 200 1407.781
   0:     0xffff94b3dee8 - <unknown>
   1:     0xffff94a7e8c0 - <unknown>
   2:     0xffff94b1bb50 - <unknown>
   3:     0xffff94b3f0ec - <unknown>
   4:     0xffff94b3ea44 - <unknown>
   5:     0xffff94b3fa04 - <unknown>
   6:     0xffff94b3f3ec - <unknown>
   7:     0xffff94b3f384 - <unknown>
   8:     0xffff94b3f378 - <unknown>
   9:     0xffff945e5730 - <unknown>
  10:     0xffff948b4404 - <unknown>
  11:     0xffff94788cd8 - <unknown>
  12:     0xffff946292ac - <unknown>
  13:     0xffff9464eae8 - <unknown>
  14:     0xffff94b4058c - <unknown>
  15:     0xffff9520ee58 - <unknown>
  16:     0xffff95277f9c - <unknown>
  17:                0x0 - <unknown>
[ERROR] Unexpected exit from worker-1
[INFO] Shutting down granian
```

Unfortunately, that's all. Happy to provide more info with some guidance.

--- a/src/runtime.rs
+++ b/src/runtime.rs
@@ -204,7 +204,11 @@ where
                     Python::with_gil(|_| drop(aw));
                 });
             },
-            () = cancel_tx.notified() => {}
+            () = cancel_tx.notified() => {
+                let _ = rb.run(move || {
+                    Python::with_gil(|_| drop(aw));
+                });
+            }
         }
     });
 
@@ -249,7 +253,14 @@ where
                     });
                 });
             },
-            () = cancel_tx.notified() => {}
+            () = cancel_tx.notified() => {
+                let _ = rb.run(move || {
+                    Python::with_gil(|py| {
+                        drop(fut_ref);
+                        drop(event_loop_ref);
+                    });
+                });
+            }
         }
     });
 

